### ISSUE TITLE
Clarify type checking in statement stacks

### ISSUE THREAD

**Issue by Gauntlet173** [google/blockly#4609]
**Where**

https://developers.google.com/blockly/guides/create-custom-blocks/type-checks

**What**

  
- [ x] Text
- [ ] Image or Gif
- [ ] Other

**Old content**

There are three places where typed statements are defined: on the previous and next connections of a block, and on a statement input's connection. These define stacking notches on the top and bottom of a block that can only be connected to blocks whose notches are either untyped, or whose notches contain at least one matching type.

The following example demonstrates "Mouse" and "Rat" types, where either type of block can be added to a statement input RODENTS. However, this toy example does not support both mice and rats in the rodents list.

**Suggested content**

There are three places where typed statements are defined: on the previous and next connections of a block, and on a statement input's connection. These define stacking notches on the top and bottom of a block that can only be connected to blocks whose notches are either untyped, or whose notches contain at least one matching type.

In a statement stack, a statement block will successfully connect to the bottom of another statement block if any type allowed for `previousConnection` in the bottom statement block is also allowed for `nextConnection` in the top statement block. It will also successfully connect if the `previousConnection` value is null. If you are connecting a statement block to a statement input on a parent block, it will successfully connect if any type allowed for `previousConnection` is also allowed in the `check` value of the statement input, or if the `previousConnection` value of the statement is null. The `check` property of a statement input is only used to check the type of the first block in the stack connected to it.

The following example demonstrates "Mouse" and "Rat" types, where either type of block can be added to a statement input RODENTS. However, this toy example does not support both mice and rats in the rodents list.

**Additional context**

Hopefully this is helpful, but please check it for correctness, particularly with regard to nulls. It may also be the case that connections succeed if the Check or nextStatement is set to null, I'm not sure. Or I may have gotten something wrong.

**Comment by maribethb** [comment]
Thanks for the suggestion! I played around with the type checking and verified that `null` can be in any of the 3 type locations and still connect any block (e.g. if `nextConnection` is null, any block will connect regardless of that block's `previousConnection` value). I've added that information to what you wrote here.

I've started a change to our documentation site, and will update you here when it's been published.
